### ISSUE TITLE
Potential for further spatial optimization of generated CDXJ

### ISSUE THREAD

**Issue by machawk1** [oduwsdl/ipwb#228]
There exists the potential for further de-duplication of CDXJ entries, particularly from multiple capture sources. Two WARCs, with entries containing headerT1/payloadX and headerT2/payloadX could potentially be merged, as the payload/response body is the same.

This might also have ramifications of squashing important information that resides in the headers that cause browsers to behave differently on replay based on which header is served.

This should be investigated further and provided as an option.

Credit to @b5 for suggestion.

**Comment by ibnesayeed** [comment]
Wait, what do you mean by merging here? Entries in the CDXJ file are observations (their key is SURT+datetime), so no deduplication is desired there, which, if done, would be lossy. As far as content (payload) deduplication is concerned, it will happen automatically by IPFS. This is actually one of the core benefits of utilizing IPFS. It doesn't matter where and how `PayloadX` was observed, be it the same URL retrieved again and again from the same host, or many different URLs returning the same payload, observed from on client or multiple clients at different times, nothing really matters, the content will be deduplicated automatically. Every observation will have an entry in respective index file and when multiple indexes are merged, nothing should be squashed. Am I missing something here?

**Comment by machawk1** [comment]
```
com,example)/ 20140203120001 {"locator": "urn:ipfs/Qm..00/Qm..AA", "mime_type": "text/html", "status_code": "200"}
com,example)/ 20140204120003 {"locator": "urn:ipfs/Qm..01/Qm..AA", "mime_type": "text/html", "status_code": "200"}
com,example)/ 20140205120051 {"locator": "urn:ipfs/Qm..03/Qm..AA", "mime_type": "text/html", "status_code": "200"}
...
com,example)/ 20160203170001 {"locator": "urn:ipfs/Qm..0n/Qm..AA", "mime_type": "text/html", "status_code": "200"}
```

Should we store index entries for the sake of differing HTTP headers over time if the payload/entity body never changes?

**Comment by ibnesayeed** [comment]
> Should we store index entries for the sake of differing HTTP headers over time if the payload/entity body never changes?

Yes!

It is a precious information for non-transactional archives. As long as SURT+datetime (and any other key info kept outside the JSON block) produces unique records, we should keep that. Squashing unchanged entries would be lossy. Here is an example to describe the danger. Suppose a resource was observed every single day for last several years, but the payload never changed till date. We will have thousands of entries in the index, but they all are strong proof that the content didn't change at over observation times over a long period. Now, suppose we squash every entry except the first, now the users would assume we have only observed the resource once, a few years ago, so we have essentially lost the info we gathered. Similarly, preserving the last entry would limit us in expressing when the resource was observed first. If we store first and last entries, then also we will lose the ability to bring confidence that the content was the same in the interim. There are ways to express the confidence during the replay time that the content was not changed in certain windows of time, but that can only be done if the underlying index data is preserved and not squashed. We can even allow a query option for the client to ask for a TimeMap that only lists the URI-Ms with changed content. OWB exposes this information in its tabular view by putting asterisks when a changes were observed.

**Comment by machawk1** [comment]
@ibnesayeed The purpose of this ticket was to explore the potential and the ramifications. Thanks for highlighting these. Closing this ticket for now. Ping @b5 re:@ibnesayeed's explanation.

**Comment by ibnesayeed** [comment]
I used a term `non-transactional archives` in the last post, which, for the sake of context, I think I should put some more details about. It is a term that is used in contrast to systems that provide exact time when a change occurred, such as Wiki or version management systems. A similar analogy can be found in the time-series data, whether it is a an event-based entry or a periodic observation. In case of transactional archives, such as Wiki, any number of observations between two consecutive changes do not provide any extra information. In contrast, non-transactional archives are external observers that do not know the state of the resource over time at moments other than when observed. Hence, more frequent observations (revisits) increase the probability of catching more changes in the interim, but there is no guarantee that the resource was not changed during two consecutive observations that yield the same payload. Decreasing the frequency of observations further degrade the confidence. It is possible that the resource was changed a few times after the first observation, then reverted to the previous state before the next observation.

/cc @b5
